client/my-sites/term-tree-selector/add-term.jsx

```diff
@@ -0,0 +1,232 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes, Component } from 'react';
+import ReactDom from 'react-dom';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import { localize } from 'i18n-calypso';
+import { find } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Dialog from 'components/dialog';
+import TermTreeSelectorTerms from './terms';
+import Button from 'components/button';
+import Gridicon from 'components/gridicon';
+import FormInputValidation from 'components/forms/form-input-validation';
+import FormTextInput from 'components/forms/form-text-input';
+import FormSectionHeading from 'components/forms/form-section-heading';
+import FormCheckbox from 'components/forms/form-checkbox';
+import FormLabel from 'components/forms/form-label';
+import FormLegend from 'components/forms/form-legend';
+import FormFieldset from 'components/forms/form-fieldset';
+import viewport from 'lib/viewport';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getPostTypeTaxonomy } from 'state/post-types/taxonomies/selectors';
+import { getTerms } from 'state/terms/selectors';
+import { addTerm } from 'state/terms/actions';
+
+class TermSelectorAddTerm extends Component {
+	static initialState = () => {
+		return Object.assign( {}, {
+			showDialog: false,
+			selectedParent: [],
+			isTopLevel: true,
+			isValid: false,
+			error: null
+		} );
+	};
+
+	static propTypes = {
+		labels: PropTypes.object,
+		postType: PropTypes.string,
+		siteId: PropTypes.number,
+		terms: PropTypes.array,
+		taxonomy: PropTypes.string,
+		translate: PropTypes.func
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = this.constructor.initialState();
+		this.boundCloseDialog = this.closeDialog.bind( this );
+		this.boundOpenDialog = this.openDialog.bind( this );
+		this.boundOnParentChange = this.onParentChange.bind( this );
+		this.boundOnSearch = this.onSearch.bind( this );
+		this.boundSaveTerm = this.saveTerm.bind( this );
+		this.boundOnTopLevelChange = this.onTopLevelChange.bind( this );
+		this.boundValidateInput = this.validateInput.bind( this );
+	}
+
+	onSearch( searchTerm ) {
+		this.setState( { searchTerm: searchTerm } );
+	}
+
+	openDialog( event ) {
+		event.preventDefault();
+
+		this.setState( {
+			showDialog: true,
+			selectedParent: []
+		} );
+	}
+
+	closeDialog() {
+		this.setState( this.constructor.initialState() );
+	}
+
+	onParentChange( item ) {
+		this.setState( {
+			selectedParent: [ item.ID ],
+			isTopLevel: false
+		}, this.isValid );
+	}
+
+	onTopLevelChange() {
+		this.setState( {
+			isTopLevel: ! this.state.isTopLevel,
+			selectedParent: []
+		}, this.isValid );
+	}
+
+	getSelectedValues() {
+		const termName = ReactDom.findDOMNode( this.refs.termName ).value.trim();
```

Comment:
Should we track `termName` in state by changing the `<FormTextInput />` into a [controlled component](https://facebook.github.io/react/docs/forms.html#controlled-components) to avoid the `ref` and `findDOMNode`? Both work fine, but if for example we were to change `<FormTextInput />` to be wrapped in a div, this would no longer work.